api/server/httputils/write_log_stream.go

```diff
@@ -17,11 +18,15 @@ import (
 // WriteLogStream writes an encoded byte stream of log messages from the
 // messages channel, multiplexing them with a stdcopy.Writer if mux is true
 func WriteLogStream(_ context.Context, w io.Writer, msgs <-chan *backend.LogMessage, config *container.LogsOptions, mux bool) {
+	// Used instead of the Flush() we had previously, so we don't set the header twice because of the otel wrapper
+	// see here: https://github.com/moby/moby/issues/47448
+	if rw, ok := w.(http.ResponseWriter); ok {
+		rw.WriteHeader(http.StatusOK)
```

Comment:
Trying to figure out how writing this here would prevent OTEL from trying to write a header again; wouldn't it still hit the "already wrote header" case?


https://github.com/moby/moby/blob/f9dfd139ec0d169cf0cf6b534fad7d98137fcc23/vendor/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/wrap.go#L93-L99

Reply:
> Trying to figure out how writing this here would prevent OTEL from trying to write a header again; wouldn't it still hit the "already wrote header" case?
> 
> https://github.com/moby/moby/blob/f9dfd139ec0d169cf0cf6b534fad7d98137fcc23/vendor/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/wrap.go#L93-L99


Our Flusher calls `Flush()` on the stdlib implementation, not on the OTEL wrapper because it doesn't wrap the `Flush()` func.   WriteHeader(), so if we call this here we are setting `w.wroteHeader` to true on the OTEL wrapper, so that subsequent calls to the OTEL wrapper's `Write()` func will NOT call the OTEL wrapper's `WriteHeader()` func.  

---

example of current sitation: 

call to `flusher.Flush()`;
- stdlib `Flush(`) gets called by `flusher.Flush()`;
- stdlib `Write()` gets called by stdlib `Flush()`;
- stdlib `WriteHeader()` gets called by stdlib`Write()`

call to flusher.Write():
- OTEL wrapper `Write()` gets called by `flusher.Write()`;
- OTEL wrapper `WriteHeader()` gets called - only when it hasn't been called before;
- stdlib `WriteHeader()` gets called by otel wrapper `WriteHeader()`;
- stdlib `Write()` gets called by otel wrapper `Write()`;
- `flusher.Flush()` gets called by `flusher.Write()`;
- ... repeat above flow 


Sorry for the verbosity, too much wrapping confusion here :)

Reply:
It seems like we either to need to manually instrument the streaming responses or provide a smarter writer implementation to the daemon so that the first flush calls WriteHeader explicitly.